pageserver/src/metrics.rs

```diff
@@ -1272,6 +1284,85 @@ impl Drop for SmgrOpTimer {
         if let Some(per_timeline_getpage_histo) = &self.per_timeline_latency_histo {
             per_timeline_getpage_histo.observe(elapsed);
         }
+
+        now
+    }
+}
+
+impl SmgrOpTimer {
+    pub(crate) fn observe_smgr_op_completion_and_start_flushing(mut self) -> SmgrOpFlushInProgress {
+        let flush_start = self.smgr_op_end();
+        let SmgrOpTimerInner {
+            global_flush_in_progress_micros,
+            per_timeline_flush_in_progress_micros,
+            ..
+        } = self.0.take().unwrap();
+        SmgrOpFlushInProgress {
+            base: flush_start,
+            global_micros: global_flush_in_progress_micros.unwrap(),
+            per_timeline_micros: per_timeline_flush_in_progress_micros.unwrap(),
+        }
+    }
+}
+
+impl std::ops::Deref for SmgrOpTimer {
+    type Target = SmgrOpTimerInner;
+
+    fn deref(&self) -> &Self::Target {
+        self.0.as_ref().unwrap()
+    }
+}
+
+impl std::ops::DerefMut for SmgrOpTimer {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.0.as_mut().unwrap()
+    }
+}
+
+impl Drop for SmgrOpTimer {
+    fn drop(&mut self) {
+        if let Some(inner) = self.0.take() {
+            inner.smgr_op_end();
+            // no flush timer
+        } else {
+            // observe_smgr_op_completion_and_start_flushing was called
+        }
+    }
+}
+
+impl SmgrOpFlushInProgress {
+    pub(crate) async fn measure<Fut, O>(mut self, mut fut: Fut) -> O
+    where
+        Fut: std::future::Future<Output = O>,
+    {
+        let mut fut = std::pin::pin!(fut);
+
+        let now = Instant::now();
+        // Whenever observe_guard gets called, or dropped,
+        // it adds the time elapsed since its last call to metrics.
+        // Last call is tracked in `now`.
+        let mut observe_guard = scopeguard::guard(
+            || {
+                let elapsed = now - self.base;
+                self.global_micros
+                    .inc_by(u64::try_from(elapsed.as_micros()).unwrap());
+                self.per_timeline_micros
+                    .inc_by(u64::try_from(elapsed.as_micros()).unwrap());
+                self.base = now;
+            },
+            |mut observe| {
+                observe();
+            },
+        );
```

Comment:
nit: this is a bit too smart for my taste. You could've done it with a select, but the comment helps.